bitcoin/crawler.go

```diff
@@ -0,0 +1,539 @@
+		Agent:               bitcoinResult.Agent,
+		Protocols:           bitcoinResult.Protocols,
+		ConnectError:        bitcoinResult.ConnectError,
+		ConnectErrorStr:     bitcoinResult.ConnectErrorStr,
+		CrawlError:          bitcoinResult.Error,
+		CrawlErrorStr:       bitcoinResult.ErrorStr,
+		CrawlEndTime:        time.Now(),
+		ConnectStartTime:    bitcoinResult.ConnectStartTime,
+		ConnectEndTime:      bitcoinResult.ConnectEndTime,
+		Properties:          data,
+		LogErrors:           c.cfg.LogErrors,
+	}
+
+	// We've now crawled this peer, so increment
+	c.crawledPeers++
+
+	return cr, nil
+}
+
+func (c *Crawler) PeerProperties(node *AddrInfo) map[string]any {
+	// TODO: to be implemented later
+	properties := map[string]any{}
+
+	properties["Services"] = "TBI"
+
+	return properties
+}
+
+type BitcoinResult struct {
+	ConnectStartTime time.Time
+	ConnectEndTime   time.Time
+	ConnectError     error
+	ConnectErrorStr  string
+	Agent            string
+	ProtocolVersion  int32
+	Protocols        []string
+	ListenAddrs      []ma.Multiaddr
+	Error            error
+	ErrorStr         string
+	RoutingTable     *core.RoutingTable[PeerInfo]
+}
+
+func (c *Crawler) crawlBitcoin(ctx context.Context, pi PeerInfo) chan BitcoinResult {
+	resultCh := make(chan BitcoinResult)
+
+	go func() {
+		result := BitcoinResult{}
+		neighbours := make([]PeerInfo, 0)
+
+		addrs := pi.Addrs()
+
+		// addrInfo := peer.AddrInfo{
+		// 	ID:    pi.ID(),
+		// 	Addrs: sanitizedAddrs,
+		// }
+
+		var conn net.Conn
+		var err error
+		connectionMaxRetry := 10
+		result.ConnectStartTime = time.Now()
+		if len(addrs) > 0 {
+			for i := 0; i < connectionMaxRetry; i++ {
+				netAddr, _ := manet.ToNetAddr(addrs[0])
+				d := net.Dialer{
+					Timeout: c.cfg.DialTimeout,
+				}
+				conn, err = d.DialContext(ctx, netAddr.Network(), netAddr.String())
+				if err == nil {
+					break
+				}
+			}
+		}
+
+		if conn == nil {
+			result.RoutingTable = &core.RoutingTable[PeerInfo]{
+				PeerID:    pi.ID(),
+				Neighbors: neighbours,
+				ErrorBits: uint16(0), // FIXME
+				Error:     result.Error,
+			}
+			result.Error = err
+			resultCh <- result
+			close(resultCh)
+			return
+		}
+		// conn, result.ConnectError = c.connect(ctx, addrInfo) // use filtered addr list
+		result.ConnectEndTime = time.Now()
+
+		// If we could successfully connect to the peer we actually crawl it.
+		if result.ConnectError == nil {
+
+			nodeRes, err := c.Handshake(conn)
+			result.Agent = nodeRes.UserAgent
+			result.ProtocolVersion = nodeRes.ProtocolVersion
+			if err != nil {
+				log.Debugf("[%s] Handshake failed: %v", addrs, err)
+			}
+
+			err = c.WriteMessage(conn, wire.NewMsgGetAddr())
+			if err != nil {
+				log.Warningf("[%s] GetAddr failed: %v", addrs, err)
+			}
+
+			firstReceived := -1
+			tolerateMessages := 5
+			// The nodes send a lot of inv messages
+			tolerateInvMessages := 50
+			tolerateVerAckMessages := 10
+			toleratePingMessages := 3
```

Comment:
could you in a brief code comment explain the logic behind the message "toleration"